dateutil/parser.py

```diff
@@ -602,8 +602,6 @@ def parse(self, timestr, default=None, ignoretz=False, tzinfos=None, **kwargs):
                     raise ValueError("Offset must be tzinfo subclass, "
                                      "tz string, or int offset.")
                 ret = ret.replace(tzinfo=tzinfo)
-            elif res.tzname and res.tzname in time.tzname:
```

Comment:
I think I meant more in the abstract, I didn't really think much about the implementation. Really, though, I think a better fix is going to be somewhat more complicated, as I mentioned in #318. Basically, I think what should be done is something like this (eliding over the details because I haven't though through the logic fully):

```python
# Code to determine if it's supposed to be `tzlocal()`
if isinstance(ret.tzinfo, tz.tzlocal):
    if ret.tzname() != res.tzname or not tz.datetime_exists(ret):
        if tz.datetime_ambiguous(ret) and tz.enfold(ret, fold=1).tzname() == res.tzname:
            ret = tz.enfold(ret, fold=1)
        elif res.tzoffset == 0:
            ret = ret.replace(tzinfo=tz.tzutc())
        elif res.tzoffset:
            ret = ret.replace(tzinfo=tz.tzoffset(res.tzname, res.tzoffset))
```

It probably needs to be a bit more complicated than that to handle all the edge cases properly, and to disambiguate ambiguous dates, but the logic is at least somewhat clear:

1. If it could be parsed as `tzlocal()` but applying `tzlocal()` to the date doesn't give you the input `tzname` you expected (likely because of a DST mismatch), check if it's just an ambiguous date.
2. If it's an ambiguous date, keep `tzlocal()` and set the `fold` attribute to 1 to disambiguate per PEP-495.
3. If it's not an ambiguous date, try all the other possible ways to parse the date that aren't `tzlocal()`, and if you find a good one, apply it.
4. If you didn't find a good one, that's an unpleasant situation, but to maintain backwards compatibility, just leave `tzlocal()` attached (we can add a `FutureWarning` if need be on this).

I'm thinking the best implementation would be to instead of using `if / elif`, after the `tzinfos` part has been parsed, just start a queue of possible time zones, so like in this case the queue would be something like `[tzlocal(), `tzutc()`], then assign the first one in the list that creates a non-imaginary date, and if none of them create non-imaginary dates, return the first one in the list.